### PR TITLE
refactor: add error handling for Bytecode type

### LINKED ISSUE #378: Better error handling for `EvmCode` to `Bytecode` conversion
https://github.com/risechain/pevm/blob/bed6f220c343d03e6bef0dc7540c531444450d57/src/storage.rs#L106-L107

### PR DESCRIPTION
**Pull request by TropicalDog17**
Fixes #378

### CODE REVIEW

**Comment by TropicalDog17** [comment]
\cc @hai-rise

**Comment by TropicalDog17** [comment]
Seem the tests is falling, I will try to fix asap @hai-rise

**Comment by hai-rise** [comment]
@TropicalDog17 Yes, conversions in the main library must also handle the new `Result`. Thank you for working on this  🙏!

**Comment by TropicalDog17** [comment]
Hi @hai-rise, seems CI is passing now, can you take a look :)

File: src/storage.rs [hunk 1]
```diff
@@ -189,28 +209,30 @@ impl<'a, S: Storage> DatabaseRef for StorageWrapper<'a, S> {
...
+        }))
     }
 
     fn code_by_hash_ref(&self, code_hash: B256) -> Result<Bytecode, Self::Error> {
-        self.0
-            .code_by_hash(&code_hash)
-            .map(|code| code.map(Bytecode::from).unwrap_or_default())
+        self.0.code_by_hash(&code_hash).map(|code_option| {
+            code_option
+                .map(|c| Bytecode::try_from(c).unwrap_or_default())
```
Comment (hai-rise) [inline src/storage.rs:233]: This falls back to a default bytecode on conversion error which seems quite dangerous. I suppose we'd need something like this:
```rust
enum StorageWrapperError<S: Storage> {
    StorageError(S::Error),
    InvalidBytecode(BytecodeConversionError),
}
```

Then `type Error = StorageWrapperError<S>;` for proper error-handling.

File: src/storage.rs [hunk 2]
```diff
@@ -189,28 +209,30 @@ impl<'a, S: Storage> DatabaseRef for StorageWrapper<'a, S> {
...
     fn basic_ref(&self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
-        Ok(if let Some(basic) = self.0.basic(&address)? {
-            let code_hash = self.0.code_hash(&address)?;
-            let code = if let Some(code_hash) = &code_hash {
-                self.0.code_by_hash(code_hash)?.map(Bytecode::from)
-            } else {
-                None
-            };
-            Some(AccountInfo {
+        let basic = self.0.basic(&address)?;
```
Comment (hai-rise) [inline src/storage.rs:212]: Perhaps just flatten it to this?
```rust
let Some(basic) = self
    .0
    .basic(&address)
    .map_err(StorageWrapperError::StorageError)?
else {
    return Ok(None);
};
```

**Review by hai-rise — CHANGES_REQUESTED** [review]
I've left some comments. Some parts can be quite confusing. We should definitely improve error handling going forward!

File: Cargo.toml [hunk 3]
```diff
@@ -57,6 +57,7 @@ serde_json = { version = "1.0.128", features = ["preserve_order"] }
 op-alloy-consensus = { version = "0.3.3", optional = true }
 op-alloy-network = { version = "0.3.3", optional = true }
 op-alloy-rpc-types = { version = "0.3.3", optional = true }
+thiserror = "1.0.64"
```
Comment (hai-rise) [inline Cargo.toml:60]: Let's put this under `smallvec`.

File: src/storage.rs [hunk 4]
```diff
@@ -180,49 +194,77 @@ pub trait Storage {
     fn block_hash(&self, number: &u64) -> Result<B256, Self::Error>;
 }
 
+/// Errors when wrapping storage.
```
Comment (hai-rise) [inline src/storage.rs:197]: `/// revm [Database] errors when using pevm [Storage] as the underlying provider.`

File: src/storage.rs [hunk 5]
```diff
@@ -180,49 +194,77 @@ pub trait Storage {
...
-                balance: basic.balance,
-                nonce: basic.nonce,
-                code_hash: code_hash.unwrap_or(KECCAK_EMPTY),
-                code,
+        self.0
+            .basic(&address)
+            .map_err(StorageWrapperError::StorageError)
+            .map(|basic_option| {
+                basic_option.map(|basic| {
+                    let code_hash = self.0.code_hash(&address).ok().flatten();
```
Comment (hai-rise) [inline src/storage.rs:220]: Storage errors should be returned instead of defaulting to a `None` here.

File: src/storage.rs [hunk 6]
```diff
@@ -180,49 +194,77 @@ pub trait Storage {
...
+                    let code = code_hash
+                        .and_then(|hash| self.0.code_by_hash(&hash).ok().flatten())
```
Comment (hai-rise) [inline src/storage.rs:222]: Storage errors should be returned instead of defaulting to a `None` here.

File: src/storage.rs [hunk 7]
```diff
@@ -180,49 +194,77 @@ pub trait Storage {
...
+                        .map(|c| {
+                            Bytecode::try_from(c)
+                                .map_err(|e| Self::Error::InvalidBytecode(e))
+                                .unwrap_or_default()
```
Comment (hai-rise) [inline src/storage.rs:226]: We must return the error, not defaulting after mapping it!

File: src/vm.rs [hunk 8]
```diff
@@ -98,6 +99,8 @@ pub enum ReadError {
     /// The stored memory value type doesn't match its location type.
     /// TODO: Handle this at the type level?
     InvalidMemoryValueType,
+    /// The bytecode is invalid and cannot be converted.
```
Comment (hai-rise) [inline src/vm.rs:102]: Nit but can you put this above `InvalidMemoryValueType`?

**Review by hai-rise — CHANGES_REQUESTED** [review]
More nits :pray:.

File: src/storage.rs [hunk 9]
```diff
@@ -303,14 +325,26 @@ mod tests {
...
     fn eof_bytecodes() {
         let bytecode = Bytecode::Eof(Arc::new(Eof::decode(EOF_BYTECODE).unwrap()));
         let evm_code = EvmCode::from(bytecode.clone());
         assert!(eq_bytecodes(&bytecode, &evm_code));
-        assert_eq!(bytecode, Bytecode::from(evm_code));
+        assert_eq!(bytecode, evm_code.try_into().unwrap());
+    }
+    #[test]
+    fn eof_bytecodes_error() {
+        let dangling_data = bytes!("010203").to_vec();
```
[the remark this answers is not in the document]
Comment (hai-rise) [reply src/storage.rs:366]: @TropicalDog17 You may have missed this as it was hidden by GitHub!

**Review by hai-rise — CHANGES_REQUESTED** [review]
Almost there!

File: src/storage.rs [hunk 10]
```diff
@@ -303,14 +356,30 @@ mod tests {
...
+    #[test]
+    fn eof_bytecodes_error() {
+        assert_eq!(
+            Bytecode::try_from(EvmCode::Eof(Bytes::new())),
+            Err(BytecodeConversionError::EofDecodingError)
+        );
+
+        let dangling_data = bytes!("010203").to_vec();
+        let mut eof_dangling = EOF_BYTECODE.to_vec();
+        eof_dangling.extend(dangling_data);
```
[the remark this answers is not in the document]
Comment (TropicalDog17) [reply src/storage.rs:379]: yeah pretty sure I've fixed this

**Review by hai-rise — APPROVED** [review]
Thank you for your patience! Hopefully, #382 will also remove most of these complications in the future.

**Comment by TropicalDog17** [comment]
> Thank you for your patience! Hopefully, #382 will also remove most of these complications in the future.

Yeah I'm also very appreciated your detailed review @hai-rise

**Pull request merged**